### ISSUE TITLE
bug: Qwen3.6-35B-A3B (MoE) gives different temp-0 output across fresh server processes for the same prompt — stable within-process, diverges cross-process

### ISSUE THREAD

**Issue by Pushkinist** [Pushkinist/rMLX#445]
## Observation

`Qwen3.6-35B-A3B` (`mlx-community/Qwen3.6-35B-A3B-8bit`, resolves to `Qwen3_5MoeForConditionalGeneration`) produced two different temp-0 outputs for the identical 4k-token prompt across two different `rmlx serve` processes (observed as `c86f04a9…` vs `dc6d30ad…`). Each process's output was stable across repeated runs *within* that process — the divergence only appears when the server is freshly relaunched.

Conditions: `temp 0`, `--kv-quant none`, `--max-ctx 16384`, 200 tokens generated, streamed SSE, 4k-token prompt.

Dense `Qwen3.8-27B` and `Gemma4` were process-identical under the same harness at the same conditions — so this is not a harness/decoding-loop bug, it is specific to (at least) this model. Throughput was unaffected. Not yet investigated beyond this observation — no logits were dumped, no bisection across layers was done.

## What "Gemma4" covers here is not pinned down

The observation doesn't say whether the "process-identical" Gemma4 run was the dense e4b/e2b snapshot or `gemma-4-26b-a4b-it-mxfp8`, which is **also MoE** (`switch_mlp`, same router shape as Qwen3.5/3.6-MoE — see `crates/rmlx-models/src/gemma4/layers/moe.rs`). If it was the a4b MoE variant, "this is MoE-specific" is falsified and the cause narrows to something Qwen3.5/3.6-MoE-specific (GatedDeltaNet, or the arch's speculative/GDN kernel path) rather than MoE routing in general. **Whoever reproduces this needs to pin which Gemma4 snapshot was actually run.**

## Static reading — credible suspects

Read-only, no execution (another agent held the GPU). Findings below, ranked by supporting evidence.

### 1. MoE router top-k tie-break via `argpartition` — leading hypothesis, unconfirmed

`crates/rmlx-models/src/qwen3_5_moe/moe.rs:238` (`SparseMoeBlock::forward`):
```rust
let part_idx = argpartition(&gates, -(tk as i32), -1, device)?; // [n, num_experts]
```
selects the top-`k` experts from `softmax(gate_logits)`. `argpartition` is a thin FFI wrapper directly over `mlx_argpartition_axis` (`crates/rmlx-mlx/src/ops/arith.rs:292`) — its tie-break rule (which index wins when two expert gate scores are numerically equal or within float noise) is entirely MLX/mlx-c internal, not rMLX code.

This repo has direct, closed precedent for exactly this failure shape: **#373** ("the host sampler and the GPU argmax disagree at temperature ~0 on Qwen3-dense — tie-break, not a bad read", merged as `01bc01b3`). That investigation proved MLX's GPU **argmax** resolves exact ties to the **lowest index**, deterministically, backed by a mutation-tested GPU test. But:
- That fix only touched the final-logits sampler path (`sampler.rs`), not `argpartition`.
- **No equivalent determinism test exists for `argpartition`'s tie-break anywhere in this repo** (`grep -rn "argpartition" crates/rmlx-mlx` shows only the FFI wrapper and its `Result` plumbing, no dedicated tie-break test the way `mlx_argmax_breaks_ties_to_lowest_index_gpu` covers argmax).
- The router turns a *continuous* near-tie into a *discrete* different-expert selection every decode step. Dense architectures (Qwen3.8-27B, dense Gemma4) have no equivalent discrete branch point mid-layer, so equivalent-magnitude float noise there stays smooth instead of flipping a hard branch — consistent with dense models being process-identical and MoE not.
- I could not determine from rMLX code alone whether MLX's internal reduction order for `argpartition` (or the softmax/gate matmul feeding it) can vary by Metal command-queue/context in a way that is fixed for the life of one process but differs across fresh process launches — that would require either reading MLX's own kernel source (not vendored in this repo) or instrumenting per-token router `expert_idx` output across two live processes.

**This is the most architecturally consistent explanation but is not proven.**

### 2. SSD KV tier — checked, most likely NOT engaged under the stated conditions

- The tier is **off by default**: `kv_ssd_cache_gb: f64` defaults to `0.0` (`crates/rmlx-cli/src/main.rs:656`), and the stated repro conditions don't include `--kv-ssd-cache-gb`.
- Per `docs/SSD_TIER.md:28-30`: "The tier is off by default... When off, the spill and hydrate hooks are never installed and decode is byte-identical to the RAM-only path."
- Both `compute_layout_key` (`crates/rmlx-kv-ssd/src/ssd_tier.rs:229-257`) and `KvQuant::cache_key_salt` (`crates/rmlx-kv-quant/src/quant.rs:740-755`) are pure FNV-1a-64 hashes over static, deterministic config (arch string, layer-quant vector, `n_kv_heads`, `head_dim`, `kv_quant` Display string) — **no process-random seed anywhere in either**, confirmed by reading both functions in full and their determinism unit tests (`compute_layout_key_is_deterministic`, `cache_key_salt_is_unique_and_deterministic`).
- Unless the observing harness passed `--kv-ssd-cache-gb` (not stated), this subsystem is inert here and should be deprioritized as a suspect. It is not fully ruled out only because the exact CLI invocation used for the observation wasn't captured.

### 3. GatedDeltaNet (GDN) linear-attention block — Qwen3.5/3.6-MoE-unique, but structurally designed to be exact

Qwen3.5/3.6-MoE is the only tested arch with a `GatedDeltaNet` block (`crates/rmlx-models/src/qwen3_5_moe/gated_delta_net.rs`, `crates/rmlx-models/src/gated_delta_msl.rs`, `crates/rmlx-models/src/metal/gated_delta_step.metal`) — absent from dense Qwen3.8 and (at least the dense variant of) Gemma4. It's a chunked-recurrence op, which is the classic shape for reduction-order sensitivity.

However, reading `gated_delta_msl.rs:265-266`, the kernel dispatch is a fixed grid `(32, dv, batch*hv)` with threadgroup `(32, 4, 1)`, and the per-timestep recurrence loop runs **sequentially within a single thread** with the state carried in registers (comment at `gated_delta_net.rs:228-243` states this is designed to be numerically exact across chunk boundaries, matching mlx-lm's kernel path). This is a parallel-across-(head,dim), sequential-within-thread design — not a cross-thread reduction — so it's structurally a weaker candidate than the router tie-break. Flagging it because it's the single largest architectural difference for this model class, not because I found a concrete defect in it.

## What a reproduction needs

1. The exact CLI invocation used for the original observation (was `--kv-ssd-cache-gb` set? was `--project` used? any speculative/MTP flags?), and which Gemma4 snapshot (dense vs `26b-a4b` MoE) was the comparison point.
2. Two fresh `rmlx serve` processes on `Qwen3.6-35B-A3B-8bit`, same prompt, same flags, `--log verbose` (or a temporary instrumentation) to dump the router's `expert_idx` / top-2 gate scores per MoE layer per token, and compare token-by-token across the two processes to find the first divergent layer/token.
3. If the divergence traces to the router: dump the raw top-2 `gates` values at the first divergent step in both processes and check bit-equality, mirroring the diagnostic recipe #373 used for the sampler-level tie (`docs/SAMPLING.md`'s census method is directly reusable).
4. A determinism test for `argpartition`'s tie-break, analogous to `mlx_argmax_breaks_ties_to_lowest_index_gpu`, would close the biggest gap identified above regardless of what the reproduction finds.

## Not investigated / explicitly unknown

- Whether MLX's `argpartition`/softmax-reduction order can vary across Metal contexts/processes at all — I could not verify this against MLX's own kernel source, which is not vendored in this repo.
- Whether the SSD tier was actually enabled in the original observation's harness.
- Which Gemma4 snapshot served as the "process-identical" comparison.
